src/api/encode.rs

```diff
@@ -0,0 +1,175 @@
+// LNP/BP Rust Library
+// Written in 2020 by
+//     Dr. Maxim Orlovsky <orlovsky@pandoracore.com>
+// Encoding strategies concept developed by
+//     Martin Habovštiak <martin.habovstiak@gmail.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the MIT License
+// along with this software.
+// If not, see <https://opensource.org/licenses/MIT>.
+
+
+use std::vec::IntoIter;
+use std::convert::TryFrom;
+use zmq::Message;
+
+use bitcoin::consensus::encode::{
+    serialize as consensus_serialize,
+    deserialize as consensus_deserialize
+};
+
+use crate::bp::ShortId;
+#[cfg(feature="use-rgb")]
+use crate::csv;
+use crate::csv::{network_serialize, network_deserialize};
+use super::{Multipart, Error};
+
+
+// 1. Encoding messages
+pub trait MessageEncode where Self: Sized {
+    type Error: std::error::Error;
+    fn into_message(self) -> Message;
+    fn try_from_message(message: Message) -> Result<Self, Self::Error>;
+}
+
+// This is a trick for rust compiler helping to distinguish types implementing
+// mutually-exclusive traits (required until negative trait impls will be there)
+// Implemented after concept by Martin Habovštiak <martin.habovstiak@gmail.com>
+mod strategy {
+    use core::marker::PhantomData;
+
+    // Defining strategies:
+    pub enum BitcoinConsensus { }
+    #[cfg(feature="use-rgb")]
+    pub enum RGBStrategy { }
+    pub enum Native { }
+    pub trait Other { type Strategy; }
+
+    #[repr(transparent)]
+    pub struct Holder<T, S>(T, PhantomData<S>);
+    impl<T, S> Holder<T, S> {
+        pub fn new(val: T) -> Self { Self(val, PhantomData::<S>::default()) }
+        pub fn into_inner(self) -> T { self.0 }
+    }
+}
+
+// 1.1. Auto impl for bitcoin-serialized types
+impl<T> MessageEncode for strategy::Holder<T, strategy::BitcoinConsensus>
+    where T: bitcoin::consensus::encode::Encodable +
+             bitcoin::consensus::encode::Decodable {
+    type Error = Error;
+    fn into_message(self) -> Message {
+        Message::from(consensus_serialize(&self.into_inner()))
+    }
+    fn try_from_message(message: Message) -> Result<Self, Self::Error> {
```

Comment:
Forgot to add in the other comment that `&Message` might be more useful as well.

Reply:
If we can use `to_` instead of `into_` for passing references; is there some convention for `from_`s passing references and not consuming the argument?

Reply:
Also, another important point: we use `impl From` to actually serialize many data structures into Message, and this is just a helper method. If we will pass a reference here, we will have to consume the argument still with the underlying `From`, so I don't understand the reason of changing `self->&self` in `into/from_message` functions here...

Reply:
I'm not aware of any convention for from. I've seen `from_*` function take values and references. Just `to_*`/`into_*` are differentiated. I suppose because `Foo::from_something(&bar)` immediately tells the reader what's going on, where `bar.to_foo()` wouldn't if not for the convention.

`From` could be implemented for references too. Implementing it for references only might be even better as it might help avoiding accidental clones.

Reply:
Tried to move from `self`s to `&self`, however got plenty of problems throughout the code; in particular many parts depend on further `into_/from_` consuming `self`s, so I will keep it this way for now and at some point later, when I will have more time, revisit this and refactor the most of codebase